rfcs/002 - Custom Module Functions RFC.md

```diff
@@ -0,0 +1,80 @@
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Luisfer De Pombo <luisfer@iasql.com>
+- Alejandro Guillen <alejandro@iasql.com>
+- Yolanda Robla <yolanda@iasql.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- David Ellis <david@iasql.com>
+
+## Summary
+
+There are situations where the best (or only) way a module can provide a good experience for the service in question is via a custom Postgres function that is under-the-hood bound to behavior within the `iasql-engine`. Examples of this include querying DynamoDB/Redis/ElasticSearch/etc from within IaSQL, constructing a Lambda function payload zip from raw sources defined in SQL queries, executing an existing lambda function within a SQL query and operating on its resulting payload, etc.
+
+A mechanism to write custom functions in a generalized way across modules is proposed that will also simplify the logic within the [`scheduler.ts`](../src/services/scheduler.ts) file, along with a proposed Module interface change to shore up an issue that has recurred where custom functions get accidentally dropped from usage when module entity migration generation occurs.
+
+## Proposal
+
+The [`iasql_functions` module](../src/modules/0.0.14/iasql_functions) provides various `iasql_*` functions that under-the-hood are powered by a job processing system defined in [`scheduler.ts`](../src/services/scheduler.ts). This module and the scheduler are uncomfortably coupled via the [`enum` defined in the `iasql_functions` entity](../src/modules/0.0.14/iasql_functions/entity/index.ts) to link [these SQL functions](../src/modules/0.0.14/iasql_functions/sql/create_fns.sql) to [the actual functions](../src/services/iasql.ts) that do the work in question.
+
+This has always been a sore point architecturally as encapsulation of information is broken, and it is not truly possible to adjust the behavior of the functions in question without affecting prior module versions at the same time. It is also very difficult (though not impossible) to introduce new functions because 4 different files across the repository must be updated in sync to do so correctly, and can only be done in a way that would be considered a SemVer minor update, major is completely not allowed.
+
+It is proposed to deprecate/drop the `iasql_operation` entity and associated enum, and create a new graphile worker job type, called `rpc` and a new entity named `iasql_rpc` with no enum. Instead two fields would be added as an "addressing system" of module name and method name, which the module would implement within an `rpc` sub-object where all properties are the method names and the values are async functions that take a `Context` object as the first argument and then any arguments from the SQL side are added as follow-up parameters, all as strings similar to the current `params` array.
```

Comment:
So on module installation the `rpc` object will be read and make all the necessary inserts into the `iasql_rpc` table automatically for us?

Reply:
So there shouldn't be any need to insert anything into the `iasql_rpc` table. What I meant was that it would auto-generate plpgsql functions to be added after all migration/sql commands have been run on install and similar delete commands on uninstall.

These would use a generic plpgsql function kinda like [`iasql_cloud_manipulation`](https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.14/iasql_functions/sql/create_fns.sql#L75-L96) that is given metadata about the RPC function to call and returns the output in a generalized way that can then be consumed in a mostly-automatic way on the other side.

Reply:
so the `iasql_rpc` table would still have an "operation" id and the output and error columns?

Reply:
I think what I'm not seeing is when the fn similar to `iasql_cloud_manipulation` will be called

Reply:
When the user invokes the RPC function in Postgres it would internally call the function that actually implements the job logic and wait for the response, while the job on the engine side would use the metadata provided to it to actually invoke the function implemented by the module.

Reply:
I think I follow it now!!!